ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java

```diff
@@ -2162,4 +2177,134 @@ static class ResourceProperty {
+      }
+      return new ArrayList<>(partitionToResourceMap.get(partitionName));
+    }
+  }
+
+  /**
+   * An implementation of {@link ResourceIdentifier} to return the resource name given by the client.
+   */
+  public static class ResourceNameIdentifier implements ResourceIdentifier {
+    private String resourceName;
+
+    /**
+     * Constructor to create a {@link ResourceNameIdentifier}.
+     * @param resourceName
+     */
+    public ResourceNameIdentifier(String resourceName) {
+      this.resourceName = resourceName;
+    }
+
+    /**
+     * Check if the given resource is valid, and then return it back.
+     * @param helixClusterManager
+     * @return
+     */
+    @Override
+    public List<String> identifyResources(HelixClusterManager helixClusterManager) {
+      if (!helixClusterManager.dcToResourceNameToTag.get(helixClusterManager.clusterMapConfig).contains(resourceName)) {
+        throw new IllegalArgumentException("Resource " + resourceName + " doesn't exist");
+      }
+      return Collections.singletonList(resourceName);
+    }
+  }
+
+  /**
+   * An implementation to return all the resource this cluster map has.
+   */
+  public static class AllResourceIdentifier implements ResourceIdentifier {
+    public static final AllResourceIdentifier DEFAULT = new AllResourceIdentifier();
+
+    /**
+     * Return all the resource names this cluster map has.
+     * @param helixClusterManager
+     * @return
+     */
+    @Override
+    public List<String> identifyResources(HelixClusterManager helixClusterManager) {
+      return new ArrayList<>(
+          helixClusterManager.dcToResourceNameToTag.get(helixClusterManager.clusterMapConfig).keySet());
+    }
+  }
+
+  /**
+   * Query the list of the resource names identified by the {@link ResourceIdentifier} and return a list of {@link ResourceInfo}.
+   * @param identifier
+   * @return
+   */
+  public List<ResourceInfo> queryResourceInfos(ResourceIdentifier identifier) {
+    List<String> resourceNames = identifier.identifyResources(this);
+    return resourceNames.stream().map(this::getResourceInfo).collect(Collectors.toList());
+  }
+
+  private ResourceInfo getResourceInfo(String resourceName) {
+    String dcName = clusterMapConfig.clusterMapDatacenterName;
+    List<String> liveInstances = new ArrayList<>(getLiveInstanceCount(resourceName));
+    List<String> unavailableInstances = new ArrayList<>(getUnavailableInstanceCount(resourceName));
+    long totalCapacity = getResourceTotalRegisteredHostDiskCapacity(resourceName);
+    long liveCapacity = getResourceAvailableRegisteredHostDiskCapacity(resourceName);
+    long unavailableCapacity = totalCapacity - liveCapacity;
+    int numPartitions = getNumberOfPartitionsInResource(resourceName);
+    int replicationFactor = 3; // by default it is 3;
+    String numReplicaStr = getResourceConfig(resourceName, dcName).getNumReplica();
+    if (numReplicaStr != null) {
```

Comment:
We may hit exception if the string is not a number.

Reply:
then we will get an exception and return 503 back to client